lib/puppet/functions/docker_params_changed.rb

```diff
@@ -0,0 +1,184 @@
+                    else
+                      "docker start #{name}"
+                    end
+    _stdout, _stderr, _status = Open3.capture3(start_command)
+    # logger.puts("### BEGIN start_container\n\n")
+    # logger.puts("command: #{start_command}\nstdout: #{stdout}\nstderr: #{stderr}\nstatus: #{status}\n\n")
+    # logger.puts("### END start_container\n\n")
+  end
+
+  def stop_container(name, osfamily, _logger = nil)
+    stop_command = if osfamily == 'windows'
+                     run_with_powershell("docker stop #{name}")
+                   else
+                     "docker stop #{name}"
+                   end
+    _stdout, _stderr, _status = Open3.capture3(stop_command)
+    # logger.puts("### BEGIN stop_container\n\n")
+    # logger.puts("command: #{stop_command}\nstdout: #{stdout}\nstderr: #{stderr}\nstatus: #{status}\n\n")
+    # logger.puts("### END stop_container\n\n")
+  end
+
+  def remove_container(name, osfamily, stop_wait_time, cidfile, logger = nil)
+    stop_command = if osfamily == 'windows'
+                     run_with_powershell("docker stop --time=#{stop_wait_time} #{name}")
+                   else
+                     "docker stop --time=#{stop_wait_time} #{name}"
+                   end
+    _stdout, _stderr, _status = Open3.capture3(stop_command)
+    # logger.puts("### BEGIN remove_container\n\n")
+    # logger.puts("command: #{stop_command}\nstdout: #{stdout}\nstderr: #{stderr}\nstatus: #{status}\n\n")
+
+    remove_command = if osfamily == 'windows'
+                       run_with_powershell("docker rm -v #{name}")
+                     else
+                       "docker rm -v #{name}"
+                     end
+    _stdout, _stderr, _status = Open3.capture3(remove_command)
+    # logger.puts("command: #{remove_command}\nstdout: #{stdout}\nstderr: #{stderr}\nstatus: #{status}\n\n")
+
+    remove_cidfile(cidfile, osfamily, logger)
+    # logger.puts("### END remove_container\n\n")
+  end
+
+  def create_container(cmd, osfamily, image, _logger = nil)
+    # logger.puts("### BEGIN create_container\n\n")
+    pull_command = if osfamily == 'windows'
+                     run_with_powershell("docker pull #{image} -q")
+                   else
+                     "docker pull #{image} -q"
+                   end
+    _stdout, _stderr, _status = Open3.capture3(pull_command)
+    # logger.puts("command: #{pull_command}\nstdout: #{stdout}\nstderr: #{stderr}\nstatus: #{status}\n\n")
+    create_command = if osfamily == 'windows'
+                       run_with_powershell(cmd)
+                     else
+                       cmd
+                     end
+    _stdout, _stderr, _status = Open3.capture3(create_command)
+    # logger.puts("command: #{create_command}\nstdout: #{stdout}\nstderr: #{stderr}\nstatus: #{status}\n\n")
+    # logger.puts("### END create_container\n\n")
+  end
+
+  def detect_changes(opts) # rubocop:disable Metrics/CyclomaticComplexity,Metrics/PerceivedComplexity
```

Comment:
This method is quite big - it might be good to refactor some of the code out in to separate methods or potentially create an object to handle the `docker inspect` output and allow us to query what we want from it a bit easier. E.g.
```ruby
class DockerInspectInspector
    def initialize(opts, docker_inspect_output)
      @opts = opts
      @docker_inspect_output = JSON.parse(docker_inspect_output)[0]
      @opt_volumes_count = @opts['volumes'].scan(%r{(?=:)}).count
      @os_family_windows = opts['osfamily'] == 'windows'
      @opts_volumes_valid = @opts['volumes'].is_a? String
    end

    def mounts_changed
      return unless @opts_volumes_valid
      @opts['volumes'].include?(':') && @opts['volumes'] != @docker_inspect_output['Mounts'].to_a[0]
    end

    def volumes_changed
      return unless @opts_volumes_valid
      !@opts['volumes'].include?(':') && @opts['volumes'] != @docker_inspect_output['Config']['Volumes'].to_a[0]
    end

    def volume_count_equal(i)
      return unless @opts_volumes_valid
      @opt_volumes_count == i
    end

    def param_changed?
      return true if mounts_changed && !@os_family_windows
      return true if volumes_changed && !@os_family_windows
      return true if volume_count_equal(2) && volumes_changed && !@os_family_windows
      volume_count_equal(1) && volumes_changed && @os_family_windows
    end

  end
```

...and then we can call the methods within the class to achieve what we want:
```ruby
stdout, stderr, status = Open3.capture3("docker inspect #{opts['sanitised_title']}")
docker_inspect_inspector = DockerInspectInspector.new(opts, stdout)

param_changed = docker_inspect_inspector.params_changed?
```

Reply:
That class is a very rough idea - do let me know if you want to go down that road before implementing. It is likely not the best way to do it, but hopefully it serves as a potential idea

Reply:
That's a great idea. I have thought to split into multiple functions, but using OOP seems to be a more readable solution.
I think this implementation will not remain in this state, it will be changed using resource API as @DavidS's recommendation.
I think all improvements will be implemented in a further ticket.
There are many improvements to be done: splitting this big function into small functions(using OOP), using ruby-pwsh,, using journalctl for logging and many other small improvements.